blindsign/blindrsa/common.go

```diff
@@ -0,0 +1,121 @@
+package blindrsa
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/sha512"
+	"crypto/subtle"
+	"errors"
+	"hash"
+	"io"
+	"math/big"
+
+	"github.com/cloudflare/circl/blindsign/blindrsa/internal/keys"
+)
+
+// ConvertHashFunction converts a crypto.Hash function to an equivalent hash.Hash type.
+func ConvertHashFunction(hash crypto.Hash) hash.Hash {
+	switch hash {
+	case crypto.SHA256:
+		return sha256.New()
+	case crypto.SHA384:
+		return sha512.New384()
+	case crypto.SHA512:
+		return sha512.New()
+	default:
+		panic(errUnsupportedHashFunction)
+	}
+}
+
+// EncodeMessageEMSAPSS hashes the input message and then encodes it using PSS encoding.
+func EncodeMessageEMSAPSS(message []byte, N *big.Int, hash hash.Hash, salt []byte) ([]byte, error) {
+	hash.Reset() // Ensure the hash state is cleared
+	hash.Write(message)
+	digest := hash.Sum(nil)
+	hash.Reset()
+	emBits := N.BitLen() - 1
+	encodedMsg, err := emsaPSSEncode(digest[:], emBits, salt, hash)
+	return encodedMsg, err
+}
+
+// GenerateBlindingFactor generates a blinding factor and its multiplicative inverse
+// to use for RSA blinding.
+func GenerateBlindingFactor(random io.Reader, N *big.Int) (*big.Int, *big.Int, error) {
+	randReader := random
+	if randReader == nil {
+		randReader = rand.Reader
+	}
+	r, err := rand.Int(randReader, N)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if r.Sign() == 0 {
+		r = bigOne
+	}
+	rInv := new(big.Int).ModInverse(r, N)
+	if rInv == nil {
+		return nil, nil, ErrInvalidBlind
+	}
+
+	return r, rInv, nil
+}
+
+// VerifyMessageSignature verifies the input message signature against the expected public key
+func VerifyMessageSignature(message, signature []byte, saltLength int, pk *keys.BigPublicKey, hash crypto.Hash) error {
```

Comment:
there are some exported functions that require keys from `/internal/keys` package, which is an internal package.
Thus, these keys cannot be instantiated in other packages (unless providing a generator function).

Can we reconsider to unexport, for example, `VerifyMessageSignature`, since we already have two constructors to get a Verifier?

Reply:
> Can we reconsider to unexport, for example, VerifyMessageSignature, since we already have two constructors to get a Verifier?

I'm not sure I follow this. Can you rephrase?

Reply:
Currently there are three methods to verify a signature:
1) building a `NewVerifier`, and calling its `Verify` method.
2) building a `NewDeterministicVerifier`, and calling its `Verify` method.
3) using the `VerifyMessageSignature` function.

I see redundancy on having so many functions that do the same thing (unless they **behave** differently).
It seems to me that 3) should not be exposed and moved to an internal package.

Reply:
Got it -- thanks for clarifying. I'll try to work things out such that (3) is no longer exported.